lte/gateway/c/core/oai/test/s1ap_task/test_s1ap_mme_handlers.cpp

```diff
@@ -381,5 +372,338 @@ TEST_F(S1apMmeHandlersTest, GenerateUEContextReleaseCommand) {
+  memset(&pdu, 0, sizeof(pdu));
+
+  ASSERT_EQ(RETURNok, s1ap_mme_decode_pdu(&pdu, payload));
+  ASSERT_EQ(
+      RETURNok, s1ap_mme_handle_message(state, assoc_id, stream_id, &pdu));
+
+  handle_mme_ue_id_notification(state, assoc_id);
+
+  ASSERT_EQ(state->mmeid2associd.num_elements, 1);
+
+  // Send UE context release command mimicing MME_APP
+  MessageDef* message_p;
+  message_p =
+      itti_alloc_new_message(TASK_MME_APP, S1AP_UE_CONTEXT_RELEASE_COMMAND);
+  S1AP_UE_CONTEXT_RELEASE_COMMAND(message_p).mme_ue_s1ap_id = 7;
+  S1AP_UE_CONTEXT_RELEASE_COMMAND(message_p).enb_ue_s1ap_id = 1;
+  S1AP_UE_CONTEXT_RELEASE_COMMAND(message_p).cause =
+      S1AP_SCTP_SHUTDOWN_OR_RESET;
+  ASSERT_EQ(
+      send_msg_to_task(&task_zmq_ctx_main_s1ap, TASK_S1AP, message_p),
+      RETURNok);
+
+  // Freeing pdu and payload data
+  bdestroy_wrapper(&payload);
+  ASN_STRUCT_FREE_CONTENTS_ONLY(asn_DEF_S1ap_S1AP_PDU, &pdu);
+  ASN_STRUCT_FREE_CONTENTS_ONLY(asn_DEF_S1ap_S1AP_PDU, &pdu_s1);
+
+  // Sleep to ensure that messages are received and contexts are released
+  std::this_thread::sleep_for(std::chrono::milliseconds(500));
+
+  ASSERT_EQ(state->mmeid2associd.num_elements, 0);
+}
+
+TEST_F(S1apMmeHandlersTest, HandleConnectionEstCnf) {
+  ASSERT_EQ(task_zmq_ctx_main_s1ap.ready, true);
+
+  S1ap_S1AP_PDU_t pdu_s1;
+  memset(&pdu_s1, 0, sizeof(pdu_s1));
+  ASSERT_EQ(RETURNok, generate_s1_setup_request_pdu(&pdu_s1));
+  ASSERT_EQ(
+      RETURNok, s1ap_mme_handle_message(state, assoc_id, stream_id, &pdu_s1));
+
+  uint8_t initial_ue_bytes[] = {
+      0x00, 0x0c, 0x40, 0x48, 0x00, 0x00, 0x05, 0x00, 0x08, 0x00, 0x02,
+      0x00, 0x01, 0x00, 0x1a, 0x00, 0x20, 0x1f, 0x07, 0x41, 0x71, 0x08,
+      0x09, 0x10, 0x10, 0x00, 0x00, 0x00, 0x00, 0x10, 0x02, 0xe0, 0xe0,
+      0x00, 0x04, 0x02, 0x01, 0xd0, 0x11, 0x40, 0x08, 0x04, 0x02, 0x60,
+      0x04, 0x00, 0x02, 0x1c, 0x00, 0x00, 0x43, 0x00, 0x06, 0x00, 0x00,
+      0xf1, 0x10, 0x00, 0x01, 0x00, 0x64, 0x40, 0x08, 0x00, 0x00, 0xf1,
+      0x10, 0x00, 0x00, 0x00, 0xa0, 0x00, 0x86, 0x40, 0x01, 0x30};
+
+  bstring payload;
+  payload = blk2bstr(&initial_ue_bytes, sizeof(initial_ue_bytes));
+  S1ap_S1AP_PDU_t pdu;
+  memset(&pdu, 0, sizeof(pdu));
+
+  ASSERT_EQ(RETURNok, s1ap_mme_decode_pdu(&pdu, payload));
+  ASSERT_EQ(
+      RETURNok, s1ap_mme_handle_message(state, assoc_id, stream_id, &pdu));
+
+  handle_mme_ue_id_notification(state, assoc_id);
+
+  ASSERT_EQ(state->mmeid2associd.num_elements, 1);
+
+  // Send UE connection establishment cnf mimicing MME_APP
+
+  ASSERT_EQ(send_conn_establishment_cnf(7, true, true), RETURNok);
+
+  // Freeing pdu and payload data
+  bdestroy_wrapper(&payload);
+  ASN_STRUCT_FREE_CONTENTS_ONLY(asn_DEF_S1ap_S1AP_PDU, &pdu);
+  ASN_STRUCT_FREE_CONTENTS_ONLY(asn_DEF_S1ap_S1AP_PDU, &pdu_s1);
+
+  // Sleep to ensure that messages are received and contexts are released
+  std::this_thread::sleep_for(std::chrono::milliseconds(500));
+}
+
+TEST_F(S1apMmeHandlersTest, HandleS1apErabRelCmd) {
```

Comment:
In general I think we should have expect calls for downstream and upstream tasks (e.g., expect to be called zero times). also should not we checking s1ap states at critical parts?

Reply:
@ulaskozat Added state validation for tests that was missing it, as well as expect calls.